monkey/monkey_island/cc/repository/mongo_node_repository.py

```diff
@@ -0,0 +1,87 @@
+from copy import deepcopy
+from typing import Any, MutableMapping, Sequence
+
+from pymongo import MongoClient
+
+from monkey_island.cc.models import CommunicationType, MachineID, Node
+
+from . import INodeRepository, RemovalError, RetrievalError, StorageError
+from .consts import MONGO_OBJECT_ID_KEY
+
+UPSERT_ERROR_MESSAGE = "An error occurred while attempting to upsert a node"
+SRC_FIELD_NAME = "machine_id"
+
+
+class MongoNodeRepository(INodeRepository):
+    def __init__(self, mongo_client: MongoClient):
+        self._nodes_collection = mongo_client.monkey_island.nodes
+
+    def upsert_communication(
+        self, src: MachineID, dst: MachineID, communication_type: CommunicationType
+    ):
+        try:
+            node_dict = self._nodes_collection.find_one({SRC_FIELD_NAME: src})
+        except Exception as err:
+            raise StorageError(f"{UPSERT_ERROR_MESSAGE}: {err}")
+
+        if node_dict is None:
+            updated_node = Node(machine_id=src, connections={dst: frozenset((communication_type,))})
```

Comment:
Why is the set frozen? Updating the node object and saving it is also a valid future scenario, I don't see what's the point of it being frozen

Reply:
One of our goals is that our objects should always be valid. pydantic does validation when attributes are set. If `my_obj.my_set` is defined as `PositiveInt`, an operation like `my_obj.my_set = {1,2,3}` is validated. But if the set is mutable, you could still do `my_obj.my_set.add(-1)`. Pydantic will not validate this. Making the set frozen means the user can replace the object's set (and validation will occur), and cannot mutate the set to be invalid.

Reply:
I think it's a bit of an overengineering/overkill. Modification becomes a lot more complex, since you need to deepcopy, modify and then freeze the set? For what? If you don't get a key error somewhere the mypy will complain that you assigned something that's not in the Enum.

Reply:
> For what? If you don't get a key error somewhere the mypy will complain that you assigned something that's not in the Enum.

> We also have UT's

Sure, but developers of 3rd-party plugins may not use mypy, or may ignore it. They may not right unit tests.

Reply:
Then they shouldn't complain if the code fails. I would prefer fewer constraints on this one if we end up having to deepcopy/change/cast everything. It's also more like "we need to validate this set", rather than "we need a frozen set".

Reply:
They won't complain, their users will complain.

The extra code to handle the frozen set took me maybe 2 minutes to write. If it ever prevents a bug, it could save someone hours. That seems like a worthwhile investment to me.